sycl/doc/syclcompat/README.md

```diff
@@ -0,0 +1,1288 @@
+# SYCLcompat
+
+SYCLcompat is a header-only library that intends to help developers familiar
+with other heterogeneous programming models (such as OpenMP, CUDA or HIP) to
+familiarize themselves with the SYCL programming API while porting their
+existing codes.
+Compatibility tools can also benefit from the reduced API size when converting
+legacy codebases.
+
+SYCLcompat provides:
+
+* A high-level API that provides closer semantics to other programming models,
+simplifying line by line conversions.
+* Alternative submission APIs that encapusulate SYCL-specific "queue" and
+"event" APIs for easier reference.
+* Ability to gradually introduce other SYCL concepts as the user familiarises
+themselves with the core SYCL API.
+* Clear distinction between core SYCL API and the compatibility interface via
+separate namespaces.
+
+## Notice
+
+Copyright © 2023-2023 Codeplay Software Limited. All rights reserved.
+
+Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are trademarks
+of The Khronos Group Inc.
+OpenCL(TM) is a trademark of Apple Inc. used by permission by Khronos.
+
+## Support
+
+SYCLcompat depends on specific DPCPP compiler extensions that may not be available to all the SYCL 2020 specification implementations.
+
+Specifically, this library depends on the following SYCL extensions:
+
+* [sycl_ext_oneapi_local_memory](../extensions/supported/sycl_ext_oneapi_local_memory.asciidoc)
+* [sycl_ext_oneapi_complex](../extensions/proposed/sycl_ext_oneapi_complex.asciidoc)
+* [SYCL_INTEL_free_function_queries](../extensions/experimental/sycl_ext_oneapi_free_function_queries.asciidoc)
+* [sycl_ext_oneapi_assert](../extensions/supported/sycl_ext_oneapi_assert.asciidoc)
+* [sycl_ext_oneapi_enqueue_barrier](../extensions/supported/sycl_ext_oneapi_enqueue_barrier.asciidoc)
+
+## Usage
+
+All functionality is available under the `syclcompat::` namespace, imported through the main header, `syclcompat.hpp`. Note that `syclcompat.hpp` does not import the sycl headers.
+
+``` cpp
+#include <sycl/syclcompat.hpp>
+```
+
+This document presents the public API under the [Features](#features) section, and provides a working [Sample code](#sample-code) using this library.
+Refer to those to learn to use the library.
+
+## Features
+
+### dim3
+
+SYCLcompat provides a `dim3` class akin to that of CUDA or HIP programming
+models. `dim3` encapsulates other languages iteration spaces that are
+represented with coordinate letters (x, y, z).
+
+```cpp
+namespace syclcompat {
+
+class dim3 {
+public:
+  const size_t x, y, z;
+  constexpr dim3(const sycl::range<3> &r);
```

Comment:
If SYCLomatic uses this to migrate
```
dim3 a(5, 6, 7);
```
There 2 migration options:
```
// Option1
syclcompat::dim3 a(sycl::range<3>(5, 6, 7));
// Option2
syclcompat::dim3 a(7, 6, 5);
```
I think the code of option1 is redundant. And option2 is very similar to current migration:
```
sycl::range<3> a(7, 6, 5);
```
Then the only advantage we use `syclcompat::dim3` in the migration is that we can access data member directly instead of calling method .x()/.y()/.z().

Will simplifing it to
```
struct dim3 {
size_t x, y, z;
};
```
be better?

Reply:
The key thing about `dim3` is that it is index-flipped with respect to a sycl::range. So, actually the migration of:

```
dim3 a(5, 6, 7);
```
would be:
```
syclcompat::dim3 a(5, 6, 7);
```

We agree that translating this to:
```
syclcompat::dim3 a(sycl::range<3>(7, 6, 5));
```
would be redundant, but we see two reasons to provide constructors taking a `sycl::range` (and conversion operators back to `sycl::range`):

1. Suppose a user has translated their code to `syclcompat` (either using SYCLomatic or manually), and then they want to *incrementally* migrate to standard SYCL, or interface their code with standard SYCL code. Suppose the user wants to make use of a library function which computes a `sycl::range` for their problem size, but their code uses `dim3`:

```c++
sycl::range<3> my_range = compute_3d_sycl_range(/*args*/);

syclcompat::dim3 my_dim3{my_range}; // with converting ctor
syclcompat::dim3 my_dim3{my_range[2], my_range[1], my_range[0]}; // without it

// user's syclcompat code...
```
The provided ctor is more concise & less error prone.

2. Including the definition of these ctors and conversion operators in [header code](https://github.com/intel/llvm/blob/8f00fb0ee96c9798014825043d95b1d8d981f901/sycl/include/syclcompat/dims.hpp#L35-L58) provides additional pseudo-documentation to make users aware that `dim3` and `sycl::range` are index-flipped.

In summary, we prefer to keep the additional functionality since it simplifies incremental porting to SYCL, minimizes risk of errors, and documents the `dim3` class.

Reply:
OK.

Then I think only one thing need note:
> So, actually the migration of:
dim3 a(5, 6, 7);
would be:
syclcompat::dim3 a(5, 6, 7);

In SYCLomatic, we have an assumption: cuda dim3 variables are only used in kernel config arguments.
So we choose to reverse the dimensions of cuda dim3 at where it is initialized/assigned, but not before submit kernel.
And according to https://github.com/intel/llvm/pull/9976/files#diff-d4bc9db1571b8e1285f6284bf3e6acd998684428de472754763c773a197c69a2R41, this constructor does not filp the index, so the migrated code by SYCLomatic will be:
```
syclcompat::dim3 a(7, 6, 5);
```

Reply:
Yes, we don't flip the indexes there. Currently in _SYCLcompat_, the `launch<function>` mechanism flips the indexes internally if you call it using a `dim3` object. So the migration of 
```
dim3 a(5, 6, 7);
```
would still be:
```
syclcompat::dim3 a(5, 6, 7);
```
and you would not need to flip the indexes if _SYCLomatic_ migrates kernel launches using _SYCLcompat_ functionalities.
As an example, SYCLomatic could translate:
```
kernel<<<grid, threads>>>(arg1, arg2, arg3);
```
as
```
syclcompat::launch<kernel>(grid, treads, arg1, arg2, arg3);
```

Reply:
ok